pyquil/noise.py

```diff
@@ -513,3 +513,141 @@ def add_decoherence_noise(prog, T1=30e-6, T2=30e-6, gate_time_1q=50e-9, gate_tim
+        therefore be a power of 2. The canonical shape has a separate axis for each qubit, such that
+        ``p[i,j,...,k]`` gives the estimated probability of bitstring ``ij...k``.
+    :param List[np.array] assignment_probabilities: A list of assignment probability matrices
+        per qubit. Each assignment probability matrix is expected to be of the form::
+
+            [[p00 p01]
+             [p10 p11]]
+
+    :return: ``p_corrected`` an array with as many dimensions as there are qubits that contains
+        the noisy-readout-corrected estimated probabilities for each measured bitstring, i.e.,
+        ``p[i,j,...,k]`` gives the estimated probability of bitstring ``ij...k``.
+    :rtype: np.array
+    """
+    return _apply_local_trafos(p, assignment_probabilities)
+
+
+def correct_bitstring_probs(p, assignment_probabilities):
+    """
+    Given a 2d array of corrupted bitstring probabilities (outer axis iterates over shots, inner
+    axis over bits) and a list of assignment probability matrices (one for each bit in the readout)
+    compute the corrected probabilities.
+
+    :param np.array p: An array that enumerates bitstring probabilities. When
+        flattened out ``p = [p_00...0, p_00...1, ...,p_11...1]``. The total number of elements must
+        therefore be a power of 2. The canonical shape has a separate axis for each qubit, such that
+        ``p[i,j,...,k]`` gives the estimated probability of bitstring ``ij...k``.
+    :param List[np.array] assignment_probabilities: A list of assignment probability matrices
+        per qubit. Each assignment probability matrix is expected to be of the form::
+
+            [[p00 p01]
+             [p10 p11]]
+
+    :return: ``p_corrected`` an array with as many dimensions as there are qubits that contains
+        the noisy-readout-corrected estimated probabilities for each measured bitstring, i.e.,
+        ``p[i,j,...,k]`` gives the estimated probability of bitstring ``ij...k``.
+    :rtype: np.array
+    """
+    return _apply_local_trafos(p, (np.linalg.inv(ap) for ap in assignment_probabilities))
+
+
+def bitstring_probs_to_z_moments(p):
+    """
+    Convert between bitstring probabilities and joint Z moment expectations.
+
+    :param np.array p: An array that enumerates bitstring probabilities. When
+        flattened out ``p = [p_00...0, p_00...1, ...,p_11...1]``. The total number of elements must
+        therefore be a power of 2. The canonical shape has a separate axis for each qubit, such that
+        ``p[i,j,...,k]`` gives the estimated probability of bitstring ``ij...k``.
+    :return: ``z_moments``, an np.array with one length-2 axis per qubit which contains the
+        expectations of all monomials in ``{I, Z_0, Z_1, ..., Z_{n-1}}``. The expectations of each
+        monomial can be accessed via::
+
+            <Z_0^j_0 Z_1^j_1 ... Z_m^j_m> = z_moments[j_0,j_1,...,j_m]
+
+    :rtype: np.array
+    """
```

Comment:
Ah, finally caught up to @ntezak.  Hadamard transform works here because n-qubit version gives (-1)^{\sum_{i}k_{i}n_{i}}  which gives the xor bitwise and thus overall parity of each bit-string.  👍

Reply:
:+1: